File: languages/java/app-encryption/src/main/java/com/godaddy/asherah/appencryption/SessionFactory.java
package com.godaddy.asherah.appencryption;

import java.time.Instant;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.LongAdder;
import java.util.function.Function;

import org.checkerframework.checker.index.qual.NonNegative;
import org.checkerframework.checker.nullness.qual.NonNull;
import org.json.JSONObject;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.github.benmanes.caffeine.cache.Cache;
import com.github.benmanes.caffeine.cache.Caffeine;
import com.github.benmanes.caffeine.cache.Expiry;
import com.github.benmanes.caffeine.cache.RemovalCause;
import com.godaddy.asherah.appencryption.envelope.EnvelopeEncryption;
import com.godaddy.asherah.appencryption.envelope.EnvelopeEncryptionBytesImpl;
import com.godaddy.asherah.appencryption.envelope.EnvelopeEncryptionJsonImpl;
import com.godaddy.asherah.appencryption.kms.KeyManagementService;
import com.godaddy.asherah.appencryption.kms.StaticKeyManagementServiceImpl;
import com.godaddy.asherah.appencryption.persistence.InMemoryMetastoreImpl;
import com.godaddy.asherah.appencryption.persistence.Metastore;
import com.godaddy.asherah.appencryption.utils.MetricsUtil;
import com.godaddy.asherah.appencryption.utils.SafeAutoCloseable;
import com.godaddy.asherah.crypto.CryptoPolicy;
import com.godaddy.asherah.crypto.NeverExpiredCryptoPolicy;
import com.godaddy.asherah.crypto.engine.bouncycastle.BouncyAes256GcmCrypto;
import com.godaddy.asherah.crypto.keys.SecureCryptoKeyMap;

import io.micrometer.core.instrument.Metrics;
import io.micrometer.core.instrument.config.MeterFilter;

public class SessionFactory implements SafeAutoCloseable {
  private static final Logger logger = LoggerFactory.getLogger(SessionFactory.class);

  private final String productId;
  private final String serviceId;
  private final Metastore<JSONObject> metastore;
  private final SecureCryptoKeyMap<Instant> systemKeyCache;
  private final CryptoPolicy cryptoPolicy;
  private final KeyManagementService keyManagementService;
  private final Cache<String, CachedEnvelopeEncryptionJsonImpl> sessionCache;

  public SessionFactory(
      final String productId,
      final String serviceId,
      final Metastore<JSONObject> metastore,
      final SecureCryptoKeyMap<Instant> systemKeyCache,
      final CryptoPolicy cryptoPolicy,
      final KeyManagementService keyManagementService) {
    this.productId = productId;
    this.serviceId = serviceId;
    this.metastore = metastore;
    this.systemKeyCache = systemKeyCache;
    this.cryptoPolicy = cryptoPolicy;
    this.keyManagementService = keyManagementService;

    this.sessionCache = Caffeine.newBuilder()
        .weigher((String intermediateKeyId, CachedEnvelopeEncryptionJsonImpl session) -> {
          if (session.isUsed()) {
            return 0;
          }
          return 1;
        })
        .maximumWeight(cryptoPolicy.getSessionCacheMaxSize())
        .expireAfter(new Expiry<String, CachedEnvelopeEncryptionJsonImpl>() {
          @Override
          public long expireAfterCreate(@NonNull final String key,
              @NonNull final CachedEnvelopeEncryptionJsonImpl value, final long currentTime) {
            // Always pin on create since we known it will be used
            return Long.MAX_VALUE;
          }

          @Override
          public long expireAfterRead(@NonNull final String key, @NonNull final CachedEnvelopeEncryptionJsonImpl value,
              final long currentTime, @NonNegative final long currentDuration) {
            // If we know it's still in use, don't expire it yet
            if (value.isUsed()) {
              return Long.MAX_VALUE;
            }

            // No longer in use, so now kickoff the expire timer
            return TimeUnit.MILLISECONDS.toNanos(cryptoPolicy.getSessionCacheExpireMillis());
          }

          @Override
          public long expireAfterUpdate(@NonNull final String key,
              @NonNull final CachedEnvelopeEncryptionJsonImpl value, final long currentTime,
              @NonNegative final long currentDuration) {
            // If we know it's still in use, don't expire it yet
            if (value.isUsed()) {
              return Long.MAX_VALUE;
            }

            // No longer in use, so now kickoff the expire timer
            return TimeUnit.MILLISECONDS.toNanos(cryptoPolicy.getSessionCacheExpireMillis());
          }
        })
        .removalListener(
            (String key, CachedEnvelopeEncryptionJsonImpl session, RemovalCause cause) -> {
              // actually close the real thing
              session.getEnvelopeEncryptionJsonImpl().close();
            })
        .build();
  }

  CachedEnvelopeEncryptionJsonImpl acquireShared(final Function<String, EnvelopeEncryptionJsonImpl> createSession,
      final String partitionId) {
    return sessionCache.asMap().compute(partitionId, (k, cachedSession) -> {
      if (cachedSession == null) {
        // Creating for first time and increment usage counter as we're the first user
        CachedEnvelopeEncryptionJsonImpl newSession =
            new CachedEnvelopeEncryptionJsonImpl(createSession.apply(partitionId), partitionId);
        newSession.incrementUsageTracker();

        return newSession;
      }

      // Already exists in cache, so just increment usage counter
      cachedSession.incrementUsageTracker();
      return cachedSession;
    });
  }

  void releaseShared(final String partitionId) {
    // Decrements the usage counter if still in the cache
    sessionCache.asMap().computeIfPresent(partitionId, (k, cachedSession) -> {
      cachedSession.decrementUsageTracker();
      return cachedSession;
    });
  }

  public Session<JSONObject, byte[]> getSessionJson(final String partitionId) {
    EnvelopeEncryption<byte[]> envelopeEncryption = getEnvelopeEncryptionBytes(partitionId);

    return new SessionJsonImpl<>(envelopeEncryption);
  }

  public Session<byte[], byte[]> getSessionBytes(final String partitionId) {
    EnvelopeEncryption<byte[]> envelopeEncryption = getEnvelopeEncryptionBytes(partitionId);

    return new SessionBytesImpl<>(envelopeEncryption);
  }

  public Session<JSONObject, JSONObject> getSessionJsonAsJson(final String partitionId) {
    EnvelopeEncryption<JSONObject> envelopeEncryption = getEnvelopeEncryptionJson(partitionId);

    return new SessionJsonImpl<>(envelopeEncryption);
  }

  public Session<byte[], JSONObject> getSessionBytesAsJson(final String partitionId) {
    EnvelopeEncryption<JSONObject> envelopeEncryption = getEnvelopeEncryptionJson(partitionId);

    return new SessionBytesImpl<>(envelopeEncryption);
  }

  EnvelopeEncryption<byte[]> getEnvelopeEncryptionBytes(final String partitionId) {
    return new EnvelopeEncryptionBytesImpl(getEnvelopeEncryptionJson(partitionId));
  }

  private EnvelopeEncryption<JSONObject> getEnvelopeEncryptionJson(final String partitionId) {
    Function<String, EnvelopeEncryptionJsonImpl> createFunc = id -> {
      Partition partition = getPartition(partitionId);

      return new EnvelopeEncryptionJsonImpl(
          partition,
          metastore,
          systemKeyCache,
          new SecureCryptoKeyMap<>(cryptoPolicy.getRevokeCheckPeriodMillis()),
          new BouncyAes256GcmCrypto(),
          cryptoPolicy,
          keyManagementService);
    };

    if (cryptoPolicy.canCacheSessions()) {
      return acquireShared(createFunc, partitionId);
    }

    return createFunc.apply(partitionId);
  }

  Partition getPartition(final String partitionId) {
    return new Partition(partitionId, serviceId, productId);
  }

  @Override
  public void close() {
    try {
      systemKeyCache.close();
    }
    catch (Exception e) {
      logger.error("unexpected exception during skCache close", e);
    }

    // This should force everything to be evicted and process the cleanup
    sessionCache.invalidateAll();
    sessionCache.cleanUp();
  }

  public static MetastoreStep newBuilder(final String productId, final String serviceId) {
    return new Builder(productId, serviceId);
  }

  public static final class Builder implements MetastoreStep, CryptoPolicyStep, KeyManagementServiceStep, BuildStep {
    private final String productId;
    private final String serviceId;

    private Metastore<JSONObject> metastore;
    private CryptoPolicy cryptoPolicy;
    private KeyManagementService keyManagementService;
    private boolean metricsEnabled = false;

    private Builder(final String productId, final String serviceId) {
      this.productId = productId;
      this.serviceId = serviceId;
    }

    @Override
    public CryptoPolicyStep withInMemoryMetastore() {
      this.metastore = new InMemoryMetastoreImpl<>();
      return this;
    }

    @Override
    public CryptoPolicyStep withMetastore(final Metastore<JSONObject> metastoreObject) {
      this.metastore = metastoreObject;
      return this;
    }

    @Override
    public KeyManagementServiceStep withNeverExpiredCryptoPolicy() {
      this.cryptoPolicy = new NeverExpiredCryptoPolicy();
      return this;
    }

    @Override
    public KeyManagementServiceStep withCryptoPolicy(final CryptoPolicy policy) {
      this.cryptoPolicy = policy;
      return this;
    }

    @Override
    public BuildStep withStaticKeyManagementService(final String demoMasterKey) {
      this.keyManagementService = new StaticKeyManagementServiceImpl(demoMasterKey);
      return this;
    }


    @Override
    public BuildStep withKeyManagementService(final KeyManagementService kms) {
      this.keyManagementService = kms;
      return this;
    }

    @Override
    public BuildStep withMetricsEnabled() {
      this.metricsEnabled = true;
      return this;
    }

    @Override
    public SessionFactory build() {
      if (!metricsEnabled) {
        // Deny takes precedence in the filtering logic, so we deny if they didn't explicitly enable metrics
        Metrics.globalRegistry.config().meterFilter(MeterFilter.denyNameStartsWith(MetricsUtil.AEL_METRICS_PREFIX));
      }

      return new SessionFactory(productId, serviceId, metastore,
          new SecureCryptoKeyMap<>(cryptoPolicy.getRevokeCheckPeriodMillis()), cryptoPolicy, keyManagementService);
    }
  }

  public interface MetastoreStep {
    // Leaving this here for now for user integration test convenience. Need to add "don't run in prod" checks somehow
    CryptoPolicyStep withInMemoryMetastore();

    CryptoPolicyStep withMetastore(Metastore<JSONObject> metastore);
  }

  public interface CryptoPolicyStep {
    KeyManagementServiceStep withNeverExpiredCryptoPolicy();

    KeyManagementServiceStep withCryptoPolicy(CryptoPolicy policy);
  }

  public interface KeyManagementServiceStep {
    // Leaving this here for now for user integration test convenience. Need to add "don't run in prod" checks somehow
    BuildStep withStaticKeyManagementService(String demoMasterKey);

    BuildStep withKeyManagementService(KeyManagementService keyManagementService);
  }

  public interface BuildStep {
    BuildStep withMetricsEnabled();

    SessionFactory build();
  }

  class CachedEnvelopeEncryptionJsonImpl implements EnvelopeEncryption<JSONObject> {
    private final EnvelopeEncryptionJsonImpl envelopeEncryptionJsonImpl;
    // The usageCounter is used to determine if any callers are still using this instance.
    private final LongAdder usageCounter = new LongAdder();
    private final String key;

    CachedEnvelopeEncryptionJsonImpl(final EnvelopeEncryptionJsonImpl envelopeEncryptionJsonImpl, final String key) {
      this.envelopeEncryptionJsonImpl = envelopeEncryptionJsonImpl;
      this.key = key;
    }

    @Override
    public JSONObject encryptPayload(final byte[] payload) {
      return envelopeEncryptionJsonImpl.encryptPayload(payload);
    }

    @Override
    public byte[] decryptDataRowRecord(final JSONObject dataRowRecord) {
      return envelopeEncryptionJsonImpl.decryptDataRowRecord(dataRowRecord);
    }

    @Override
    public void close() {
      releaseShared(key);
    }

    EnvelopeEncryptionJsonImpl getEnvelopeEncryptionJsonImpl() {
      return envelopeEncryptionJsonImpl;
    }

    void incrementUsageTracker() {
      usageCounter.increment();
    }

    void decrementUsageTracker() {
      usageCounter.decrement();
    }

    boolean isUsed() {
      return usageCounter.sum() > 0;
    }
  }
}

File: languages/csharp/AppEncryption/AppEncryption/SessionFactory.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Threading;
using App.Metrics;
using App.Metrics.Concurrency;
using GoDaddy.Asherah.AppEncryption.Envelope;
using GoDaddy.Asherah.AppEncryption.Kms;
using GoDaddy.Asherah.AppEncryption.Persistence;
using GoDaddy.Asherah.AppEncryption.Util;
using GoDaddy.Asherah.Crypto;
using GoDaddy.Asherah.Crypto.Engine.BouncyCastle;
using GoDaddy.Asherah.Crypto.Keys;
using GoDaddy.Asherah.Logging;
using Microsoft.Extensions.Caching.Memory;
using Microsoft.Extensions.Logging;
using Newtonsoft.Json.Linq;

namespace GoDaddy.Asherah.AppEncryption
{
    public class SessionFactory : IDisposable
    {
        #pragma warning disable SA1401
        protected internal readonly MemoryCache SessionCache;
        #pragma warning restore SA1401

        // Percentage of session cache to compact if it exceeds size limits and remove unused sessions
        private const int CompactionPercentage = 50;

        private static readonly ILogger Logger = LogManager.CreateLogger<SessionFactory>();

        private readonly string productId;
        private readonly string serviceId;
        private readonly IMetastore<JObject> metastore;
        private readonly SecureCryptoKeyDictionary<DateTimeOffset> systemKeyCache;
        private readonly CryptoPolicy cryptoPolicy;
        private readonly KeyManagementService keyManagementService;
        private readonly ConcurrentDictionary<string, object> semaphoreLocks;

        public SessionFactory(
            string productId,
            string serviceId,
            IMetastore<JObject> metastore,
            SecureCryptoKeyDictionary<DateTimeOffset> systemKeyCache,
            CryptoPolicy cryptoPolicy,
            KeyManagementService keyManagementService)
        {
            this.productId = productId;
            this.serviceId = serviceId;
            this.metastore = metastore;
            this.systemKeyCache = systemKeyCache;
            this.cryptoPolicy = cryptoPolicy;
            this.keyManagementService = keyManagementService;
            semaphoreLocks = new ConcurrentDictionary<string, object>();
            SessionCache = new MemoryCache(new MemoryCacheOptions());
        }

        public interface IMetastoreStep
        {
            // Leaving this in here for now for user integration test convenience. Need to add "don't run in prod" checks somehow
            ICryptoPolicyStep WithInMemoryMetastore();

            ICryptoPolicyStep WithMetastore(IMetastore<JObject> metastore);
        }

        public interface ICryptoPolicyStep
        {
            IKeyManagementServiceStep WithNeverExpiredCryptoPolicy();

            IKeyManagementServiceStep WithCryptoPolicy(CryptoPolicy cryptoPolicy);
        }

        public interface IKeyManagementServiceStep
        {
            // Leaving this in here for now for user integration test convenience. Need to add "don't run in prod" checks somehow
            IBuildStep WithStaticKeyManagementService(string demoMasterKey);

            IBuildStep WithKeyManagementService(KeyManagementService keyManagementService);
        }

        public interface IBuildStep
        {
            IBuildStep WithMetrics(IMetrics metrics);

            SessionFactory Build();
        }

        public static IMetastoreStep NewBuilder(string productId, string serviceId)
        {
            return new Builder(productId, serviceId);
        }

        public virtual void Dispose()
        {
            try
            {
                // only close system key cache since we invoke its creation
                systemKeyCache.Dispose();
            }
            catch (Exception e)
            {
                Logger.LogError(e, "unexpected exception during skCache close");
            }

            // Actually dispose of all the remaining sessions that might be active in the cache.
            lock (SessionCache)
            {
                foreach (KeyValuePair<string, object> sessionCacheKey in semaphoreLocks)
                {
                    CachedSession cachedSession = SessionCache.Get<CachedSession>(sessionCacheKey.Key);

                    // We need to check this to ensure that the entry was not removed by the expiration policy
                    if (cachedSession != null)
                    {
                        // actually close the real thing
                        cachedSession.GetEnvelopeEncryptionJsonImpl().Dispose();
                    }

                    // now remove the entry from the cache
                    SessionCache.Remove(sessionCacheKey.Key);
                }
            }
        }

        public Session<JObject, byte[]> GetSessionJson(string partitionId)
        {
            IEnvelopeEncryption<byte[]> envelopeEncryption = GetEnvelopeEncryptionBytes(partitionId);

            return new SessionJsonImpl<byte[]>(envelopeEncryption);
        }

        public Session<byte[], byte[]> GetSessionBytes(string partitionId)
        {
            IEnvelopeEncryption<byte[]> envelopeEncryption = GetEnvelopeEncryptionBytes(partitionId);

            return new SessionBytesImpl<byte[]>(envelopeEncryption);
        }

        public Session<JObject, JObject> GetSessionJsonAsJson(string partitionId)
        {
            IEnvelopeEncryption<JObject> envelopeEncryption = GetEnvelopeEncryptionJson(partitionId);

            return new SessionJsonImpl<JObject>(envelopeEncryption);
        }

        public Session<byte[], JObject> GetSessionBytesAsJson(string partitionId)
        {
            IEnvelopeEncryption<JObject> envelopeEncryption = GetEnvelopeEncryptionJson(partitionId);

            return new SessionBytesImpl<JObject>(envelopeEncryption);
        }

        internal IEnvelopeEncryption<byte[]> GetEnvelopeEncryptionBytes(string partitionId)
        {
            return new EnvelopeEncryptionBytesImpl(GetEnvelopeEncryptionJson(partitionId));
        }

        internal Partition GetPartition(string partitionId)
        {
            return new Partition(partitionId, serviceId, productId);
        }

        /// <summary>
        /// Atomically acquires a shared <code>CachedSession</code> from the session cache for the
        /// <code>partitionId</code>, creating a new one using the given function if needed. This is used to track the
        /// number of concurrent users so cache eviction policies don't remove an entry while it's still potentially in
        /// use.
        /// </summary>
        ///
        /// <returns>The cached session that's mapped for the given <code>partitionId</code></returns>
        ///
        /// <param name="createSessionFunc">the function to create a new session if there is no current mapping</param>
        /// <param name="partitionId">the partition id for a session</param>
        private CachedSession AcquireShared(
            Func<EnvelopeEncryptionJsonImpl> createSessionFunc, string partitionId)
        {
            object getCachedItemLock = semaphoreLocks.GetOrAdd(partitionId, new object());
            CachedSession cachedItem;

            // TryGetValue is not thread safe and hence we need a lock
            lock (getCachedItemLock)
            {
                if (!SessionCache.TryGetValue(partitionId, out cachedItem))
                {
                    // If the cache size is greater than the maximum count, compact the cache
                    // This will remove all the unused sessions
                    if (SessionCache.Count >= cryptoPolicy.GetSessionCacheMaxSize())
                    {
                        SessionCache.Compact(CompactionPercentage);
                    }

                    // Creating for first time
                    cachedItem = new CachedSession(createSessionFunc(), partitionId, this);
                    var cacheEntryOptions = new MemoryCacheEntryOptions()
                        .SetPriority(CacheItemPriority.NeverRemove);

                    // Save data in cache.
                    SessionCache.Set(partitionId, cachedItem, cacheEntryOptions);
                }

                // Increment the usage counter of the entry
                cachedItem.IncrementUsageTracker();
            }

            return cachedItem;
        }

        /// <summary>
        /// Atomically marks a shared <code>CachedSession</code> in the session cache as no longer being used by the
        /// current caller for the <code>partitionId</code>. This is used to track the number of concurrent users so
        /// cache eviction policies don't remove an entry while it's still potentially in use.
        /// </summary>
        ///
        /// <param name="partitionId">the partition id for a session</param>
        private void ReleaseShared(string partitionId)
        {
            try
            {
                CachedSession cacheItem = SessionCache.Get<CachedSession>(partitionId);

                // Decrements the usage counter of the entry
                cacheItem.DecrementUsageTracker();

                // If we know it's still in use, don't expire it yet
                if (!cacheItem.IsUsed())
                {
                    // No longer in use, so now kickoff the expire timer
                    var cacheEntryOptions = new MemoryCacheEntryOptions()
                        .SetPriority(CacheItemPriority.Low)
                        .SetSlidingExpiration(TimeSpan.FromMilliseconds(cryptoPolicy.GetSessionCacheExpireMillis()))
                        .RegisterPostEvictionCallback((id, value, reason, state) =>
                        {
                            ((CachedSession)value).GetEnvelopeEncryptionJsonImpl().Dispose();
                        });
                    SessionCache.Set(partitionId, cacheItem, cacheEntryOptions);
                }
            }
            catch (Exception e)
            {
                Logger.LogError(e, "Unexpected exception during dispose");
            }
        }

        private IEnvelopeEncryption<JObject> GetEnvelopeEncryptionJson(string partitionId)
        {
            // Wrap the creation logic in a lambda so the cache entry acquisition can create a new instance when needed
            Func<EnvelopeEncryptionJsonImpl> createSessionFunc = () =>
            {
                Partition partition = GetPartition(partitionId);

                return new EnvelopeEncryptionJsonImpl(
                    partition,
                    metastore,
                    systemKeyCache,
                    new SecureCryptoKeyDictionary<DateTimeOffset>(cryptoPolicy.GetRevokeCheckPeriodMillis()),
                    new BouncyAes256GcmCrypto(),
                    cryptoPolicy,
                    keyManagementService);
            };

            if (cryptoPolicy.CanCacheSessions())
            {
                return AcquireShared(createSessionFunc, partitionId);
            }

            return createSessionFunc();
        }

        // Calling it CachedSession but we actually cache the implementing envelope encryption class.
        private class CachedSession : IEnvelopeEncryption<JObject>
        {
            private readonly EnvelopeEncryptionJsonImpl envelopeEncryptionJsonImpl;

            // The usageCounter is used to determine if any callers are still using this instance.
            private readonly StripedLongAdder usageCounter = new StripedLongAdder();
            private readonly string key;
            private readonly SessionFactory sessionFactory;

            public CachedSession(
                EnvelopeEncryptionJsonImpl envelopeEncryptionJsonImpl,
                string key,
                SessionFactory sessionFactory)
            {
                this.envelopeEncryptionJsonImpl = envelopeEncryptionJsonImpl;
                this.key = key;
                this.sessionFactory = sessionFactory;
            }

            public void Dispose()
            {
                // Instead of closing the session, we call the release function so it can atomically update the usage counter
                sessionFactory.ReleaseShared(key);
            }

            public byte[] DecryptDataRowRecord(JObject dataRowRecord)
            {
                return envelopeEncryptionJsonImpl.DecryptDataRowRecord(dataRowRecord);
            }

            public JObject EncryptPayload(byte[] payload)
            {
                return envelopeEncryptionJsonImpl.EncryptPayload(payload);
            }

            internal void IncrementUsageTracker()
            {
                usageCounter.Increment();
            }

            internal EnvelopeEncryptionJsonImpl GetEnvelopeEncryptionJsonImpl()
            {
                return envelopeEncryptionJsonImpl;
            }

            internal void DecrementUsageTracker()
            {
                usageCounter.Decrement();
            }

            internal bool IsUsed()
            {
                return usageCounter.GetValue() > 0;
            }
        }

        private class Builder : IMetastoreStep, ICryptoPolicyStep, IKeyManagementServiceStep, IBuildStep
        {
            private readonly string productId;
            private readonly string serviceId;

            private IMetastore<JObject> metastore;
            private CryptoPolicy cryptoPolicy;
            private KeyManagementService keyManagementService;
            private IMetrics metrics;

            internal Builder(string productId, string serviceId)
            {
                this.productId = productId;
                this.serviceId = serviceId;
            }

            public ICryptoPolicyStep WithInMemoryMetastore()
            {
                metastore = new InMemoryMetastoreImpl<JObject>();
                return this;
            }

            public ICryptoPolicyStep WithMetastore(IMetastore<JObject> metastore)
            {
                this.metastore = metastore;
                return this;
            }

            public IKeyManagementServiceStep WithNeverExpiredCryptoPolicy()
            {
                cryptoPolicy = new NeverExpiredCryptoPolicy();
                return this;
            }

            public IKeyManagementServiceStep WithCryptoPolicy(CryptoPolicy policy)
            {
                cryptoPolicy = policy;
                return this;
            }

            public IBuildStep WithStaticKeyManagementService(string demoMasterKey)
            {
                keyManagementService = new StaticKeyManagementServiceImpl(demoMasterKey);
                return this;
            }

            public IBuildStep WithKeyManagementService(KeyManagementService kms)
            {
                keyManagementService = kms;
                return this;
            }

            public IBuildStep WithMetrics(IMetrics metrics)
            {
                this.metrics = metrics;
                return this;
            }

            public SessionFactory Build()
            {
                // If no metrics provided, we just create a disabled/no-op one
                if (metrics == null)
                {
                    metrics = new MetricsBuilder()
                        .Configuration.Configure(options => options.Enabled = false)
                        .Build();
                }

                MetricsUtil.SetMetricsInstance(metrics);

                return new SessionFactory(
                    productId,
                    serviceId,
                    metastore,
                    new SecureCryptoKeyDictionary<DateTimeOffset>(cryptoPolicy.GetRevokeCheckPeriodMillis()),
                    cryptoPolicy,
                    keyManagementService);
            }
        }
    }
}
